include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,2378 @@
+    }
+    return it->second;
+  }
+
+  void get_obj_flds(const obj_id_t &id, variable_vector_t &obj_flds) const {
+    assert(m_flds_id_map);
+    for (auto kv : (*m_flds_id_map)) {
+      if (kv.second == id) {
+        obj_flds.push_back(kv.first);
+      }
+    }
+  }
+
+  void update_fields_id_map(const variable_t &rgn, const obj_id_t &id) {
+    // update m_flds_id_map
+    if (!m_flds_id_map) { // create a object fields - id map if it is null
+      m_flds_id_map =
+          std::make_shared<std::unordered_map<variable_t, obj_id_t>>();
+    }
+    auto it = (*m_flds_id_map).find(rgn);
+    if (it != (*m_flds_id_map).end()) {
+      it->second = id;
+    } else {
+      (*m_flds_id_map).insert({rgn, id});
+    }
+  }
+
+  bool is_rgn_obj_id(variable_t rgn, obj_id_t id) { return id == rgn; }
+
+  /***************** Base address operations *****************/
+  // get or create a variable representing the base address by given a reference
+  // this also works for base address of mru object.
+  // For mru object, we use a variable named `<obj_id>_mru_base`.
+  variable_t get_or_insert_base_addr(const variable_t &v) {
+    // if v is a reference variable,
+    assert(v.get_type().is_reference() || v.get_type().is_region());
+    if (!m_refs_base_addrs_map) { // create a ref - base addr map if it is null
+      m_refs_base_addrs_map =
+          std::make_shared<std::unordered_map<variable_t, variable_t>>();
+    }
+    auto it = (*m_refs_base_addrs_map).find(v);
+    if (it != (*m_refs_base_addrs_map).end()) {
+      return it->second;
+    }
+    varname_t v_var_name = v.name();
+    std::string str_base_name =
+        v.get_type().is_reference() ? "_base" : "_mru_base";
+    varname_t base_name =
+        v_var_name.get_var_factory().get(v_var_name, str_base_name);
+    variable_t v_base(base_name, crab::INT_TYPE, 32);
+    (*m_refs_base_addrs_map).insert({v, v_base});
+    return v_base;
+  }
+
+  variable_t get_base_addr_or_fail(const variable_t &v) const {
+    assert(v.get_type().is_reference() || v.get_type().is_region());
+    if (!m_refs_base_addrs_map) {
+      CRAB_ERROR(domain_name(),
+                 "::get_base_addr_or_fail: base address map is not created");
+    }
+    auto it = (*m_refs_base_addrs_map).find(v);
+    if (it == (*m_refs_base_addrs_map).end()) {
+      CRAB_ERROR(domain_name(), "::get_base_addr_or_fail: ", v, "not found");
+    }
+    return it->second;
+  }
+
+  // check whether mru object in the left state refers the same as the mru
+  // object in the right state
+  bool
+  mru_refer_same_object(const obj_id_t &id,
+                        const address_abstract_domain_t &l_addrs_dom,
+                        const address_abstract_domain_t &r_addrs_dom) const {
+    variable_t mru_base = get_base_addr_or_fail(id);
+    const usymb_t *l_term_ptr = l_addrs_dom.get_term(mru_base);
+    const usymb_t *r_term_ptr = r_addrs_dom.get_term(mru_base);
+    if (l_term_ptr && r_term_ptr) {
```

Comment:
Is this code to check if the 2 uninterpreted constants are equal? if `not(a < b)` and `not(b < a)` it could be that `a = b` but it could also be that they are not comparable.